### PR TITLE
Implement umfMemoryProviderAllocation[Split/Merge]

### PR DESCRIPTION
**Pull request by igchor**
That is needed to support jemalloc (avoid leaking memory).

@kswiecicki could you verify if this solves all the memory leaks? On my machine, it seems to work fine after those changes.

### CODE REVIEW

File: src/provider/provider_tracking.c
Comment (pbalcer) [inline src/provider/provider_tracking.c:123]: Given that we use a tree for tracking, it might be simpler to avoid preemptively splitting/merging pages, and instead handle that during inserts/removals.

File: include/umf/memory_provider_ops.h
Comment (pbalcer) [inline include/umf/memory_provider_ops.h:145]: you are making and assumption that "left" means "numerically lower address", right? I'd call it `lowPtr`, `highPtr`.

**Review by pbalcer — COMMENTED** [review]
as for concurrency, I think if we don't add a lock we are creating a footgun. I can imagine a scenario where for example, split is called concurrently twice on the same block.

So unless, we add some mechanism for exclusive ownerships of ranges of memory, I think we should add a lock.

File: src/provider/provider_tracking.c
Comment (vinser52) [reply src/provider/provider_tracking.c:123]: it might be error-prone concerning IPC handle support. I prefer to do it explicitly when a pool manager calls merge/split functions (like implemented right now).
In general, it is an interesting case, the IPC handle is created per coarse-grain allocation (returned by the memory provider) and if two coarse-grain allocations are merged what should we do with corresponding IPC handles if them were created before? I will think more about that. I know that L0 just does not support split/merge functionality but tracking memory provider should work for all cases.

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:128]: if you allocate `splitValue` just before `critnib_insert` then you do not need to `free(splitValue)` in every error cases below.

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:188]: if you allocate `mergedValue` just before `critnib_insert` then you do not need to `free(mergedValue)` in every error case below.

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:137]: I would add `TODO` comment that logger should be called instead of just `fprintf`

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:142]: I would add `TODO` comment that logger should be called instead of just `fprintf`

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:150]: I would add `TODO` comment that logger should be called instead of just `fprintf`

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:200]: I would add `TODO` comment that logger should be called instead of just `fprintf`

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:207]: I would add `TODO` comment that logger should be called instead of just `fprintf`

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:212]: I would add `TODO` comment that logger should be called instead of just `fprintf`

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:217]: I would add `TODO` comment that logger should be called instead of just `fprintf`

File: src/provider/provider_tracking.c
Comment (vinser52) [inline src/provider/provider_tracking.c:216]: How it is possible?

**Comment by ldorau** [comment]
I confirm that the memory leak described in https://github.com/oneapi-src/unified-memory-framework/issues/147 does not occur with this change, because `arena_extent_split()` succeeds now.

**Comment by kswiecicki** [comment]
> @kswiecicki could you verify if this solves all the memory leaks? On my machine, it seems to work fine after those changes.

It does, ASan no longer reports any leaks in jemalloc.

**Comment by igchor** [comment]
> as for concurrency, I think if we don't add a lock we are creating a footgun. I can imagine a scenario where for example, split is called concurrently twice on the same block.
> 
> So unless we add some mechanism for exclusive ownerships of ranges of memory, I think we should add a lock.

Hm, should only add a mutex for split/merge calls or should we also protect trackingFree? I think freeing a memory region that is being split/merged can just be UB, right?

File: src/provider/provider_tracking.c
Comment (igchor) [reply src/provider/provider_tracking.c:216]: Well, someone (implementer of a pool) can call split with wrong arguments (size that does not match existing regions). It's a programmer error but the check is pretty cheap and I think quite useful.

File: src/provider/provider_tracking.c
Comment (igchor) [reply src/provider/provider_tracking.c:217]: Hm, do we really want to add "TODO"s to every fprintf call? I think we have a few without annotations already. When we will finally have a logging framework we'll do `grep fprintf` anyway I guess.

File: src/provider/provider_tracking.c
Comment (igchor) [reply src/provider/provider_tracking.c:123]: Yeah, I was thinking about this as well. Doing this lazily in removal would definitely be simpler (and probably faster) but as Sergey said, might be problematic for IPC. I think we can revisit this once we have IPC support merged.

File: include/umf/memory_provider_ops.h
Comment (ldorau) [reply include/umf/memory_provider_ops.h]: That's right :-)

**Comment by ldorau** [comment]
This PR is required by #153

**Comment by igchor** [comment]
> > as for concurrency, I think if we don't add a lock we are creating a footgun. I can imagine a scenario where for example, split is called concurrently twice on the same block.
> > So unless we add some mechanism for exclusive ownerships of ranges of memory, I think we should add a lock.
> 
> Hm, should only add a mutex for split/merge calls or should we also protect trackingFree? I think freeing a memory region that is being split/merged can just be UB, right?

I've added mutex for split/merge only for now.

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:163]: > size_t rightSize = totalSize - firstSize;

(`lowSize` and `highSize`) OR (`firstSize` and `secondSize`) OR (`leftSize` and `rightSize`) - do not mix them, please

File: src/pool/pool_jemalloc.c
Comment (bratpiorka) [inline src/pool/pool_jemalloc.c:216]: please assert that size, size_a and size_b > 0 and addr != NULL

File: src/provider/provider_tracking.c
Comment (bratpiorka) [inline src/provider/provider_tracking.c:131]: add a check for splitValue != NULL

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:165]: 1) `have duplicate` -> `have to duplicate` ?
2) `highPtr, rightValue` ? - maybe `highPtr` and `highValue` ?

File: src/provider/provider_tracking.c
Comment (bratpiorka) [inline src/provider/provider_tracking.c:128]: could we stay with the full name? During the review I had to scroll up to see what is "p". If you want to keep the code short it would be better to add an alias for "tracker = provider->hTracker"

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:167]: `highPtr, rightSize` - mixed

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:170]: `This is can only happen` -> `This can only happen`

File: include/umf/memory_provider.h
Comment (bratpiorka) [inline include/umf/memory_provider.h:164]: I see that in split/merge methods we don't want to have redundant arguments, so why we need "totalSize"? Shouldn't the totalSize be just the size of the alloc pointed by ptr? What if totalSize would be less/more than allocation?

File: include/umf/memory_provider.h
Comment (bratpiorka) [inline include/umf/memory_provider.h:179]: why we need the totalSize argument?

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:202]: handle malloc failure

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:206]: move that before the assignment

File: src/provider/provider_tracking.c
Comment (bratpiorka) [inline src/provider/provider_tracking.c:172]: note: I added an issue for that

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:216]: `lowPtr` and `leftValue` - mixed names (low and high OR left and right)

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:223]: `highPtr` and `rightValue` - mixed

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:249]: `have duplicate` -> `have to duplicate`

Comment (ldorau) [inline src/provider/provider_tracking.c:249]: `highPtr, rightValue` - mixed

File: src/provider/provider_tracking.c
Comment (bratpiorka) [inline src/provider/provider_tracking.c:199]: it would be better to do "tracker = provider->hTracker"

File: src/provider/provider_tracking.c
Comment (ldorau) [inline src/provider/provider_tracking.c:250]: splitting -> merging ?

File: src/provider/provider_tracking.c
Comment (bratpiorka) [inline src/provider/provider_tracking.c:201]: please check for "mergedValue != NULL"

File: src/provider/provider_tracking.h
Comment (bratpiorka) [inline src/provider/provider_tracking.h:47]: add an assert + return if handle == NULL

**Review by bratpiorka — CHANGES_REQUESTED** [review]

**Review by ldorau — CHANGES_REQUESTED** [review]

File: src/provider/provider_tracking.c
Comment (vinser52) [reply src/provider/provider_tracking.c:202]: Make sense, but in case of merge do we really need to allocate a new element? You can just update the left record with the new size and remove the right record. Did I miss something?

File: src/pool/pool_jemalloc.c
Comment (igchor) [reply src/pool/pool_jemalloc.c:220]: Done.

File: src/pool/pool_jemalloc.c
Comment (igchor) [reply src/pool/pool_jemalloc.c:238]: Done.

File: src/provider/provider_tracking.c
Comment (igchor) [reply src/provider/provider_tracking.c:206]: Done.

File: src/provider/provider_tracking.c
Comment (igchor) [reply src/provider/provider_tracking.c:207]: See my other answer about updating the values. Without this alloc we would have to modify the value in-place which means we would have to use atomic instructions.

File: src/provider/provider_tracking.c
Comment (igchor) [reply src/provider/provider_tracking.c]: Just removing now, since split/merge is called under the lock.

File: src/provider/provider_tracking.h
Comment (igchor) [reply src/provider/provider_tracking.h:61]: Can't return since this is done in the lib destructor anyway and we cannot handle a failure there.

Comment (bratpiorka) [reply src/provider/provider_tracking.h:61]: Ok but if handle == NULL then we crash when we try to deref it. Please just add an early return for this case.

File: include/umf/memory_provider.h
Comment (vinser52) [reply include/umf/memory_provider.h:164]: Here are my thoughts about this question:

1. We should not just copy jemalloc extent API because for me it looks jemalloc-specific. And it looks like we are not doing that.
2. Regarding `totalSize`. For me, it looks similar to `umfMemoryProviderFree(..., void *ptr, size_t size)` API that also requires passing the `size` of allocation pointed by `ptr`. But not all memory provider implementations need/use it: e.g. mmap implementation uses `size`, but L0 ignores it.
3. For me, bullet 2 means that UMF requires that the pool manager remember the size of each particular coarse-grain allocation but memory providers are not required to remember the size of each coarse-grain allocation. Since we already require to pass the `size` to `umfMemoryProviderFree` function what is the difference here? 
4. Regarding what if `totalSize != actual total size`, the same question is applicable for `umfMemoryProviderFree`. On our side we can do (at least debug) verification using the tracking memory provider - there is such information in the tracker.
5. Another question regarding input parameter validation, should the `firstSize` be a multiple of the page size?

Comment (bratpiorka) [reply include/umf/memory_provider.h:164]: ok I agree with bullet 3

File: src/provider/provider_tracking.h
Comment (igchor) [reply src/provider/provider_tracking.h:63]: Right, good point. Done.

File: include/umf/memory_provider.h
Comment (igchor) [reply include/umf/memory_provider.h:164]: Regarding 5: I think it is implied because I'm saying that the pointer needs to point to the beginning of the allocation and all allocations need to be page aligned.

**Review by bratpiorka — APPROVED** [review]

**Comment by bratpiorka** [comment]
@vinser52 @pbalcer @ldorau all I need 2nd approve to merge this PR - could you check if anything is missing here?

**Review by ldorau — APPROVED** [review]

File: src/provider/provider_tracking.h
Comment (vinser52) [inline src/provider/provider_tracking.h:30]: Global split/merge lock might kill scalability. As I understand the lock is needed only when concurrent split/merge occurs for the same memory buffer.

Comment (pbalcer) [reply src/provider/provider_tracking.h:30]: well, mmap is also fully synchronized, so I wouldn't worry that much. This shouldn't be a super common operation. I'd start with a simple mutex and then we can think about locking out ranges of memory if there's a perf problem.

File: src/provider/provider_tracking.c
Comment (pbalcer) [inline src/provider/provider_tracking.c:175]: yes, preallocation makes sense - should be doable once we have the base allocator.

**Pull request merged**
